src/uu/ls/src/url_escaper.rs

```diff
@@ -0,0 +1,145 @@
+use std::{
+    borrow::Cow,
+    convert::{TryFrom, TryInto},
+    ffi::OsStr,
+};
+
+use ascii::{AsciiChar, AsciiStr, AsciiString};
+/// this is logically enum ByteOrResult {Char(AsciiChar), Escape([AsciiChar; 2])}
+/// but we don't use that definition because rustc uses 3 bytes instead of 2 for it (even if we use a pair instead of the array)
+#[derive(Copy, Clone)]
+#[repr(C)]
+struct AsciiOrEscape(Option<AsciiChar>, AsciiChar);
```

Comment:
Much better than before. The comment also helps. The `#[repr(C)]` is not necessary though. You can test it with this code (including the enum, which is indeed 3 bytes):
```rust
use ascii::AsciiChar;

enum CharOrEscape {
    Char(AsciiChar),
    Escape(AsciiChar, AsciiChar)
}

struct AsciiOrEscape(Option<AsciiChar>, AsciiChar);

#[repr(C)]
struct AsciiOrEscapeC(Option<AsciiChar>, AsciiChar);

fn main() {
   println!("{:?}", std::mem::size_of::<CharOrEscape>());
   println!("{:?}", std::mem::size_of::<AsciiOrEscape>());
   println!("{:?}", std::mem::size_of::<AsciiOrEscapeC>());
}
```
(the compiler gives no guarantees about the size though)